### ISSUE TITLE
Set QA Operating System Limits for the App

### ISSUE THREAD

**Issue by bischoffa** [department-of-veterans-affairs/va-mobile-app#7601]
**Description:** 
As QA and Release Team, I want to set a limit on what operating systems we will test so that we more efficiently and effectively test. 

**Notes:** 
- Discussed during Sprint 0 
- Use data to determine cutoff to make business case
- Identify time QA will save
- Focus is that QA will stop testing, up to Flagship App if they want to build for it or decide to cutoff operating systems 
- Look into forced updates for old versions or Availability Framework messaging 

**ACs:** 

- [ ] Android OS limitation
- [ ] Apple OS limitation
- [ ] Process to constantly reassess cutting off operating systems from testing

**Comment by bischoffa** [comment]
Chatted with Therese today during 1:1 assigning to her for Q1 priority

**Comment by TKDickson** [comment]
Getting my thoughts organized on this, I see these steps, in somewhat this order:

**Testing OS cutoff**
* Grab data on OS version usage
* Analyze things to figure out OS testing cutoff recommendation
* Run recommendation past QAR team
* (Also past stakeholders?)
* Do device OS updates (as needed) to reflect current recommendations
* Communicate current practice to rest of team (non-QAR) -- do they _have_ to fall in line TBD

**Operationalize OS updates**
* Figure out frequency of updating current practice (likely annual & synched with fall OS releases, would line up with new OS testing - Tom's got a ticket for the new OS work)
* Document decision-making process
* Come up with way for reminders on this (ex: create ticket when wrapping up current work that goes on backlog and then pulled in at the right time of year; repeat at end of next year's process)

Investigate methods for encouragement/enforcement in app
* This is probably its own distinct Sprint 0, really. So maybe just spinning up a ticket for "do that sprint 0" that would get sequenced after this one.



For this sprint, I'll probably timebox getting started on the stuff above, then spin up reasonably-sized tickets I can grab in further sprints to keep the work going.

**Comment by TKDickson** [comment]
Relevant notes from [a slack conversation](https://adhoc.slack.com/archives/C02N65QRRB7/p1707494621326149?thread_ts=1707489348.781479&cid=C02N65QRRB7):

> Ok. Just went through on testing it and we will still be able to support our current set of devices for iOS. We just have to update our deployment target to 17 which will force every developer/QA here to upgrade to xcode 15 if they haven’t already, which might have us upgrade to the latest macOS
> 
> When we update RN 0.73 that will the be the last version of RN that support Android 5.0. React Native 0.74 will have a minimum sdk of Android 6.0 and will discontinue whatever support there is for Android devices that can’t go to Android 6.0.
> 
> Currently RN will still support our list of iOS devices, so nothing will change there.
> 
> TLDR is that this update requirement from Apple won’t change what devices are supported. It will soon though, Xcode 15 minimum version is iOS 12. We have a minimum of iOS 13. Previous version supported iOS 11 and the prior version to that supported iOS 9.
> 
> What happens to devices that can’t match our minimum version is that they will still be able to download and use our app, but it is capped at the last version available that supports their device and they just won’t receive further updates.
> 
> Thanks Dylan, that’s good to know.
> If I have questions about it, is it easy to tell when we’ve changed xcode and RN versions over time (like - the entire history of the app)? That might help explain which versions of the app we see people using
> 
> Xcode version not so much. For that one if you have access to the apple developer data to see users iOS version you could loosely tie it to this
> https://developer.apple.com/support/xcode/
> developer.apple.comdeveloper.apple.com
> 
> we currently deploy with version 16.2
> Xcode version 15 is April 29, 2024
> Xcode version 14.1 was April 25, 2023
> Xcode version 13 was April 25, 2022
> I can’t see any history for further back than that

**Comment by TKDickson** [comment]
I will still need to write up tickets for allll of the above (in my comment)

However -- there are already users stranded on app versions, see excerpt from Slack. Sleuthing that out and documenting it somewhere is definitely step 1. Because we'll need to decide what we do about super super old versions, sooner-than-later.

**Comment by TKDickson** [comment]
Summary thoughts

My position is that the mobile app team should have three categories of OS support: active, passive, and disallowed.

**Active:** Major OS versions in iOS and Android that we commit to proactively developing/testing with, and fully supporting app functionality in.
**Passive:** Major OS versions (iOS and Android) that we do not develop/test with, but users can still log into the app while using. 
**Disallowed:** Major OS versions in where users cannot log in/access app while using.

The goal is to have as few versions in the Disallowed category as possible, while not unrealistically overextending ourselves with what's included in the Active category. 

How we'd set these versions:
Active: using firebase and statcounter (or other similar resource), to find a roughly ~80% of users. Generally, but not always, that is typically current and one-back for iOS, and roughly the last 3 major versions for Android.
Disallowed: any version where we are unfortunately no longer able to support them. This could be for a mobile-specific reason (for example: the react native/xcode examples in an earlier comment) or a web-specific reason (mobile login is currently routed through the browser, which is tied to OS version on iOS; we don't have control over what's supported by those teams). 
Passive: all OS versions between the active and disallowed ranges


None of the above represents a major departure from what we've been doing - more of a formalization. Some key pieces of work remaining:
* Documenting this (especially before working with outside teams)
* Getting better insight into when/if users are currently disallowed due to browser version (see ticket #8271)
* Determining what we'd do with and for users on disallowed OS versions, once we reach that point

**Comment by DonMcCaugheyUSDS** [comment]
@TKDickson I like your breakdown of Active/Passive/Disallowed a lot, I think it's a good way to think about how we support mobile.  A quick look at Google Analytics shows that the last two major versions of iOS and three of Android exceeds the 80% active support target.

What do you think about using the term "Unsupported" instead of "Disallowed"?  I'm thinking our policy for these devices is "it might still work, but if it breaks, we're not promising to fix it".  I would then summarize our passive support story as "we're not testing these OS versions every sprint, but if it breaks, we will try to fix it".

Final question: do Apple and Google have clear policies/lists of old OS versions they support for critical security and bug fixes?  If so, I'm thinking any OS version not getting critical updates should automatically go on the disallow/unsupported list.

**Comment by DonMcCaugheyUSDS** [comment]
This sounds good.  I'm thinking about how we communicate the key points to folks who don't know and don't care about the nuances.  The bullet points I have in mind are:

Support levels:
- Active: tested with each release; will fix bugs
- Passive: not tested with each release but will fix bugs (when practical)
- Unsupported: not tested and won't fix bugs

Support limits:
- Active support for OS versions of at least 80% of veterans (last 2 major for iOS, last 3 for Android)
- Passive support for older OS versions
- No support after OS vendor or major dependency (login, react native) drops support
 
Dropping support includes end of dev tool support and/or end of security updates.

**Comment by TKDickson** [comment]
Documentation up for PR review after Don and I's conversation this morning.

**Comment by TKDickson** [comment]
#8664 and #8661 contain logical follow-up work from this ticket.
